File: 2023/d02.pl
#!/usr/bin/perl

use strict;
use warnings;
use List::Util qw(reduce);

$\ = "\n";

my $res = 0;
my %counts = ("red" => 12, "green" => 13, "blue" => 14);

LINE_LOOP: while (<>) {
    chomp;
    my ($game) = $_ =~ m/Game (\d+):/;
    my %mins = ("red" => 0, "green" => 0, "blue" => 0);

    while ($_ =~ /(\d+) (\w+)/g) {
        my ($quantity, $color) = ($1, $2);
        if ($quantity > $mins{$color}) {
            $mins{$color} = $quantity;
        }

        # Part 1.
        # if ($quantity > $counts{$color}) {
        #     next LINE_LOOP;
        # }
    }

    my $product = reduce { $a * $b } values %mins;
    $res += $product;

    # Part 1.
    # $res += $game;
}

print($res);

File: 2023/d02.py
# Python
import collections
import functools
import operator

# Local
import util

def part1(L):
    res = 0
    max_counts = { 'red': 12, 'green': 13, 'blue': 14 }

    for i, line in enumerate(L, 1):
        skip_game = False
        for setx in line[line.index(':') + 1:].strip().split(';'):
            if skip_game:
                break
            for pair in setx.strip().split(','):
                parts = [x.strip() for x in pair.split()]
                count, colour = int(parts[0]), parts[1]

                if count > max_counts[colour]:
                    skip_game = True
                    break

        if not skip_game:
            res += i
    return res

def part2(L):
    res = 0

    for idx, line in enumerate(L, 1):
        counts = collections.defaultdict(int)

        for setx in line[line.index(':') + 1:].strip().split(';'):
            for pair in setx.strip().split(','):
                parts = [x.strip() for x in pair.split()]
                count, colour = int(parts[0]), parts[1]
                counts[colour] = max(count, counts[colour])

        res += functools.reduce(operator.mul, counts.values(), 1)
    return res

if __name__ == "__main__":
    L = util.strip_stdin()
    results = util.run(L, 1, 2348, [part1])
    results += util.run(L, 2, 76008, [part2])
    util.print_table(results)
